d2ir/compile.go

```diff
@@ -92,7 +94,130 @@ func (c *compiler) compileClasses(m *Map) {
+	for _, f := range m.Fields {
+		if f.Name == "vars" && f.Map() != nil {
+			varsStack = append([]*Map{f.Map()}, varsStack...)
+		}
+		if f.Primary() != nil {
+			c.resolveSubstitutions(varsStack, f.LastRef().AST(), f.Primary())
+		}
+		if f.Map() != nil {
+			c.compileSubstitutions(f.Map(), varsStack)
+		}
+	}
+	for _, e := range m.Edges {
+		if e.Primary() != nil {
+			c.resolveSubstitutions(varsStack, e.LastRef().AST(), e.Primary())
+		}
+		if e.Map() != nil {
+			c.compileSubstitutions(e.Map(), varsStack)
+		}
+	}
+}
+
+func (c *compiler) resolveSubstitutions(varsStack []*Map, node d2ast.Node, scalar *Scalar) {
+	var subbed bool
+	var resolvedField *Field
+
+	switch s := scalar.Value.(type) {
+	case *d2ast.UnquotedString:
+		for i, box := range s.Value {
+			if box.Substitution != nil {
+				for _, vars := range varsStack {
+					resolvedField = c.resolveSubstitution(vars, box.Substitution)
+					if resolvedField != nil {
+						break
+					}
+				}
+				if resolvedField != nil {
+					if resolvedField.Composite != nil {
+						c.errorf(node, `cannot reference map variable "%s"`, strings.Join(box.Substitution.IDA(), "."))
+						return
+					}
+					// If lone and unquoted, replace with value of sub
+					if len(s.Value) == 1 {
+						scalar.Value = resolvedField.Primary().Value
+					} else {
+						s.Value[i].String = go2.Pointer(resolvedField.Primary().String())
+						subbed = true
+					}
+				} else {
+					c.errorf(node, `could not resolve variable "%s"`, strings.Join(box.Substitution.IDA(), "."))
+					return
+				}
+			}
+		}
+		if subbed {
+			s.Coalesce()
+			scalar.Value = d2ast.RawString(s.ScalarString(), false)
+		}
+	case *d2ast.DoubleQuotedString:
+		for i, box := range s.Value {
+			if box.Substitution != nil {
+				for _, vars := range varsStack {
+					resolvedField = c.resolveSubstitution(vars, box.Substitution)
+					if resolvedField != nil {
+						break
+					}
+				}
+				if resolvedField != nil {
+					if resolvedField.Composite != nil {
+						c.errorf(node, `cannot reference map variable "%s"`, strings.Join(box.Substitution.IDA(), "."))
+						return
+					}
+					s.Value[i].String = go2.Pointer(resolvedField.Primary().String())
+					subbed = true
+				} else {
+					c.errorf(node, `could not resolve variable "%s"`, strings.Join(box.Substitution.IDA(), "."))
+					return
+				}
+			}
+		}
+		if subbed {
+			s.Coalesce()
+		}
+	}
+}
+
+func (c *compiler) resolveSubstitution(vars *Map, substitution *d2ast.Substitution) *Field {
+	var resolved *Field
+	for _, p := range substitution.Path {
+		if vars == nil {
+			resolved = nil
+			break
+		}
+		r := vars.GetField(p.Unbox().ScalarString())
+		if r == nil {
+			resolved = nil
+			break
+		}
+		vars = r.Map()
+		resolved = r
+	}
+	return resolved
+}
+
+func (c *compiler) overlayVars(base, overlay *Map) {
+	vars := overlay.GetField("vars")
+	if vars == nil {
+		return
+	}
+
+	lVars := base.GetField("vars")
```

Comment:
```suggestion
	baseVars := base.GetField("vars")
```

not sure what you mean by lVars.

Reply:
oh oops, copied from elsewhere